docs/docs/code-culture.md

```diff
@@ -0,0 +1,121 @@
+# Code Culture
+
+## Overview
+
+The code culture is a set of defaults ascribed to by the Marathon developer team. Since writing software is an optimization problem, we guide our design decisions with defaults, not rules. Deviations from our defaults require justification.
+
+Our defaults are:
+
+* Code is legible on GitHub
+* Immutability over mutability
+* No smelly code
+* Fail loud and proud
+* Let it crash
+* Don't panic
+
+## Code is Legible on GitHub
+
+This default speaks to the desire to write code that can be understood without needing an IDE. Marathon is written in Scala, so our defaults are specific as such.
+
+### On Implicits
+
+Implicits are an important feature in Scala and are critical for the implementation of type-classes and DSLs.
+
+Prefer:
+
+* Explicit conversion over implicit
+* Explicit passing over implicit
+
+
+For our case, if implicits don't help improve the design of code, we don't use them. We don't use implicit conversions. We don't do multi-priority implicits. When type classes help improve the design of code, we use implicits for type classes (IE serialization logic).
+
+We prefer implicits to be used contextual values, such as execution contexts. However, we prefer to be conservative, and pass the value explicitly for things like validators, authenticators, and normalizers.
+
+### On Type Inference
+
+If you are writing some function chain that has multiple levels of inferred function parameter types, we prefer to specify the parameter types.
+
+Prefer:
+
+```
+groups.map { group: Group =>
+  group.apps
+    .groupBy(_.container.image)
+    .map { case (image: String, apps: Seq[App]) =>
+      image -> apps.length
+    }
+    }
+```
+
+Over:
+
+```
+groups.map { group =>
+  group.apps.groupBy(_.container.image).map { case(k, v) =>
+    k -> v.length
+  }
+}
+```
+
+### On Playing Code Golf
+
+Code golf is a game in which you implement some program in as few characters as possible.
+
+We prefer not to play this game when working on Marathon. Instead, we focus on removing noise (boilerplate), while preserving valuable signal pertaining to the problem at hand.
+
+### On Imports
+
+We prefer, almost always, that implicits go at the top of the file.
+
+Prefer:
+
+```
+import org.company.{Stringifier, Demuxer}
+```
+
+Over:
+
+```
+import org.company._
```

Comment:
We probably should spell out to avoid wild card imports.

Reply:
✅